File: 002/max_pairwise_product.py
import random


def max_pairwise_product_naive(numbers):
    n = len(numbers)
    max_product = 0
    for first in range(n):
        for second in range(first + 1, n):
            max_product = max(max_product, numbers[first] * numbers[second])
    return max_product


def max_pairwise_product(numbers):
    """
    Returns maximal pairwise product.

    Index1 stores index of the greatest number
    Index2 stores index of the second greatest number

    :param numbers: list of numbers
    :return: maximal product
    """
    n = len(numbers)

    index1 = 1 if numbers[1] > numbers[0] else 0
    index2 = 0 if index1 == 1 else 1

    for i in range(2, n):
        if numbers[index1] <= numbers[i]:
            index2 = index1
            index1 = i
        elif numbers[index2] <= numbers[i]:
            index2 = i

    max_product = numbers[index1] * numbers[index2]
    return max_product


def stress_test(n, m):
    while True:
        length_of_numbers = random.randint(2, n)
        print(str(length_of_numbers))
        numbers = [random.randint(1, m) for _ in range(length_of_numbers)]
        print(str(numbers))
        result1 = max_pairwise_product_naive(numbers)
        result2 = max_pairwise_product(numbers)
        if result1 == result2:
            print("OK")
        else:
            print("ERROR")
            break


def main():
    input_n = int(input())
    input_numbers = [int(x) for x in input().split()]
    print(max_pairwise_product(input_numbers))


if __name__ == '__main__':
    # main()
    stress_test(30, 100000000)

File: 002/max_pairwise_product.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int MaxPairwiseProductNaive(const std::vector<int>& numbers) {
    int max_product = 0;
    int n = numbers.size();

    for (int first = 0; first < n; ++first) {
        for (int second = first + 1; second < n; ++second) {
            max_product = std::max(max_product,  numbers[first] * numbers[second]);
        }
    }

    return max_product;
}


int MaxPairwiseProduct(const std::vector<int>& numbers) {
    int index1;
    int index2;
    int max_product;
    int n = numbers.size();

    if (numbers[1] > numbers[0]) {
        index1 = 1;
        index2 = 0;
    } else {
        index1 = 0;
        index2 = 1;
    }

    for (int i = 2; i < n; ++i) {
        if (numbers[index1] <= numbers[i]) {
            index2 = index1;
            index1 = i;
        } else if (numbers[index2] <= numbers[i]) {
            index2 = i;
        }
    }

    max_product = numbers[index1] * numbers[index2];
    return max_product;
}

void StressTest(int n, int m) {
    while(true) {
        int length_of_numbers = rand() % n + 2;
        std::cout << "length_of_numbers: " << length_of_numbers << "\n";
        std::vector<int> numbers(length_of_numbers);
        for (int i = 0; i < length_of_numbers; i++) {
            numbers[i] = rand() % m + 1;
            std::cout << numbers[i] << " ";
        }
        int result1 = MaxPairwiseProductNaive(numbers);
        int result2 = MaxPairwiseProduct(numbers);

        if (result1 == result2) {
            std::cout << "OK" << "\n";
        } else {
            std::cout << "ERROR" << "\n";
            break;
        }
    }
}

int main() {
    StressTest(30, 1000);
    return 0;
}
